vehicle/carwings.go

```diff
@@ -39,70 +46,166 @@ func NewCarWingsFromConfig(other map[string]interface{}) (api.Vehicle, error) {
 		return nil, err
 	}
 
+	if cc.User == "" || cc.Password == "" {
+		return nil, errors.New("missing credentials")
+	}
+
+	log := util.NewLogger("carwin")
+
 	v := &CarWings{
 		embed:    &cc.embed,
+		log:      log,
 		user:     cc.User,
 		password: cc.Password,
-		region:   cc.Region,
+		session:  &carwings.Session{Region: cc.Region},
 	}
 
-	v.chargeStateG = provider.NewCached(v.chargeState, cc.Cache).FloatGetter()
-	v.hvacG = provider.NewCached(v.hvacAPI, cc.Cache).InterfaceGetter()
+	v.statusG = provider.NewCached(func() (interface{}, error) {
+		return nil, v.status()
+	}, cc.Cache).InterfaceGetter()
 
 	return v, nil
 }
 
-// Init new carwings session
-func (v *CarWings) initSession() error {
-	v.session = &carwings.Session{
-		Region: v.region,
+// connectIfRequired will return ErrMustRetry if ErrNotLoggedIn error could be resolved
+func (v *CarWings) connectIfRequired(err error) error {
+	if err == carwings.ErrNotLoggedIn || err.Error() == "received status code 404" {
+		if err = v.session.Connect(v.user, v.password); err == nil {
+			err = api.ErrMustRetry
+		}
 	}
-
-	return v.session.Connect(v.user, v.password)
+	return err
 }
 
-// chargeState implements the api.Vehicle interface
-func (v *CarWings) chargeState() (float64, error) {
-	if v.session == nil {
-		if err := v.initSession(); err != nil {
-			return 0, api.ErrNotAvailable
+func (v *CarWings) status() error {
+	bs, err := v.session.BatteryStatus()
+	if err == nil {
+		if elapsed := time.Since(bs.Timestamp); elapsed > carwingsStatusExpiry {
+			// api result is stale
+			if v.refreshKey != "" {
+				return v.refreshResult()
+			}
+
+			if err = v.refreshRequest(); err != nil {
+				return err
+			}
+
+			err = api.ErrMustRetry
+		} else {
```

Comment:
Den verstehe ich noch nicht. Wenn es zum Timeout kommt, erzeuge ich beim ersten mal einen refreshRequest. Der wird irgendwann erfüllt und über refreshResult abgeholt. Im refreshResult wird der Key zurück gesetzt. 

Warum ist das dann hier nochmal notwendig? Gibts da irgendwo noch einen Fehler in der Logik? Es riecht immer denn die Logik so verteilt ist nach versteckten Fehlern...

Reply:
Beim Batteriestatus request wird der timestamp der letzten Antwort gelesen. Ist jetzt seit dem letzten requestfinished mit result false die Antwort eingetrudelt, wird die so genommen und kein neuer Finish Check gemacht, da der elapsed wert nur weniger Sekunden hat=> der Key wird nicht gelöscht ohne den Else Zweig nicht gelöscht. Nach Ablauf des Cache (15min) wir dann der Finish Check mit dem alten Key gemacht... Returned success und damit bleiben die alten Werte weiter gecached anstatt neue zu requesten. Erst beim nächsten die Durchlauf (wieder 15 min) wird dann ein neuer Status request abgesetzt.